### PR TITLE
CR-1077588 Add support for PCIe Firewall IP

### PR DESCRIPTION
**Pull request by houlz0507**

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-utils.c [hunk 1]
```diff
@@ -277,6 +278,18 @@ long xclmgmt_hot_reset(struct xclmgmt_dev *lro, bool force)
 #else
 		xclmgmt_reset_pci(lro);
 #endif
+		/*
+		 * For U50 built with 2RP flow. PLP Gate is closed after
+		 * pci hot reset. Any access to PLP IPs will hangs the host.
+		 * E.g. read/write BAR2 (DMA BAR).
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-utils.c:284]: So, a malicious VM can just hang the host by accessing these IPs right after hot reset? I thought this PCIE firewall IP is for preventing this from happening, no?

File: src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-utils.c [hunk 2]
```diff
@@ -277,6 +278,18 @@ long xclmgmt_hot_reset(struct xclmgmt_dev *lro, bool force)
...
 #endif
+		/*
+		 * For U50 built with 2RP flow. PLP Gate is closed after
+		 * pci hot reset. Any access to PLP IPs will hangs the host.
+		 * E.g. read/write BAR2 (DMA BAR). 
+		 * To be able to open PLP gate after hot reset, the first 64
+		 * bytes of config space has to be restored. 
+		 * In the meanwhile, XRT expects firewall trip instead of
+		 * hard hang if there is an unexpected access of non-exist
+		 * IPs. (E.g. Invalid access from a active VM)
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/mgmtpf/mgmt-utils.c:289]: So, I'm confused. If user access PLP IP before gate opens, do we see firewall trip or host hang?

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/pcie_firewall.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	mutex_lock(&firewall->fw_lock);
+	val = reg_rd(firewall, fwr_en_resp);
+	if (val & UNBLOCK_BIT(pf, bar)) {
+		xocl_info(&pdev->dev, "unblock pf%d, bar%d", pf, bar);
+		reg_wr(firewall, (~UNBLOCK_BIT(pf, bar)) & val, fwr_en_resp);
+	}
+	mutex_unlock(&firewall->fw_lock);
+
+	return 0;
+}
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/pcie_firewall.c:76]: Can the IP provide a single bit to unblock everything? When do we ever want to unblock one BAR but leave another blocked?

**Comment by dayeh-xilinx** [comment]
retest this please

**Comment by houlz0507** [comment]
retest this please

**Pull request merged**